AsyncDisplayKit/Layout/ASDimension.mm

```diff
@@ -55,19 +55,140 @@ BOOL ASRelativeDimensionEqualToRelativeDimension(ASRelativeDimension lhs, ASRela
+{
+  return CGSizeMake(ASRelativeDimensionResolve(relativeSize.width, autoSize.width, parentSize.width),
+                    ASRelativeDimensionResolve(relativeSize.height, autoSize.height, parentSize.height));
+}
+
+BOOL ASRelativeSizeEqualToRelativeSize(ASRelativeSize lhs, ASRelativeSize rhs)
+{
+  return ASRelativeDimensionEqualToRelativeDimension(lhs.width, rhs.width)
+  && ASRelativeDimensionEqualToRelativeDimension(lhs.height, rhs.height);
+}
+
+NSString *NSStringFromASRelativeSize(ASRelativeSize size)
+{
+  return [NSString stringWithFormat:@"{%@, %@}",
+          NSStringFromASRelativeDimension(size.width),
+          NSStringFromASRelativeDimension(size.height)];
+}
+
+#pragma mark - ASSize
+
+ASSize ASSizeMake()
+{
+  return (ASSize){
+    .width = ASRelativeDimensionAuto,
+    .height = ASRelativeDimensionAuto,
+    .minWidth = ASRelativeDimensionAuto,
+    .maxWidth = ASRelativeDimensionAuto,
+    .minHeight = ASRelativeDimensionAuto,
+    .maxHeight = ASRelativeDimensionAuto
+  };
+}
+
+ASSize ASSizeMakeFromCGSize(CGSize size)
+{
+  ASSize s = ASSizeMake();
+  s.width = ASRelativeDimensionMakeWithPoints(size.width);
+  s.height = ASRelativeDimensionMakeWithPoints(size.height);
+  return s;
+}
+
+BOOL ASSizeEqualToSize(ASSize lhs, ASSize rhs)
+{
+  return ASRelativeDimensionEqualToRelativeDimension(lhs.width, rhs.width)
+  && ASRelativeDimensionEqualToRelativeDimension(lhs.height, rhs.height)
+  && ASRelativeDimensionEqualToRelativeDimension(lhs.minWidth, rhs.minWidth)
+  && ASRelativeDimensionEqualToRelativeDimension(lhs.maxWidth, rhs.maxWidth)
+  && ASRelativeDimensionEqualToRelativeDimension(lhs.minHeight, rhs.minHeight)
+  && ASRelativeDimensionEqualToRelativeDimension(lhs.maxHeight, rhs.maxHeight);
+}
+
+NSString *NSStringFromASSize(ASSize size)
+{
+  return [NSString stringWithFormat:
+          @"<ASSize: exact=%@, min=%@, max=%@>",
+          NSStringFromASRelativeSize(ASRelativeSizeMake(size.width, size.height)),
+          NSStringFromASRelativeSize(ASRelativeSizeMake(size.minWidth, size.minHeight)),
+          NSStringFromASRelativeSize(ASRelativeSizeMake(size.maxWidth, size.maxHeight))];
+}
+
+static inline void ASSizeConstrain(CGFloat minVal, CGFloat exactVal, CGFloat maxVal, CGFloat *outMin, CGFloat *outMax)
+{
+    NSCAssert(!isnan(minVal), @"minVal must not be NaN");
+    NSCAssert(!isnan(maxVal), @"maxVal must not be NaN");
+    // Avoid use of min/max primitives since they're harder to reason
+    // about in the presence of NaN (in exactVal)
+    // Follow CSS: min overrides max overrides exact.
+
+    // Begin with the min/max range
+    *outMin = minVal;
+    *outMax = maxVal;
+    if (maxVal <= minVal) {
+        // min overrides max and exactVal is irrelevant
+        *outMax = minVal;
+        return;
+    }
+    if (isnan(exactVal)) {
+        // no exact value, so leave as a min/max range
+        return;
+    }
+    if (exactVal > maxVal) {
+        // clip to max value
+        *outMin = maxVal;
```

Comment:
Maybe this one doesn't need a log, but it might be nice…

This type of thing could be added in a follow up PR too.